akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1088,38 +1089,65 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
 
     private def sendEvent(event: T, promise: Promise[Done]): Promise[Done] = {
       onAsyncInput(event, OptionVal.Some(promise))
-      currentState.get() match {
-        case Completed ⇒ failPromiseOnComplete(promise)
-        case _         ⇒ promise
-      }
+      if (stopped) failPromiseOnComplete(promise)
+      else promise
     }
 
     // external call
     override def invokeWithFeedback(event: T): Future[Done] = {
       val promise: Promise[Done] = Promise[Done]()
-      promise.future.andThen { case _ ⇒ waitingForProcessing.remove(promise) }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+      promise.future.onComplete(_ ⇒ onFeedbackCompleted(promise))(ExecutionContexts.sameThreadExecutionContext)
+
+      if (waitingForProcessing.isEmpty) {
```

Comment:
Isn't here a race condition as well if multiple invocations happen at the same time?

Reply:
Or rather a race against deleting from `waitingFromProcessing`? It might be not-empty right now, but what if the last one is just removed in that moment?